analyses/npl/10_historical_event_timeline.md

```diff
@@ -0,0 +1,307 @@
+        data[f"event_rp{rp}"] = False
+        data[f"event_rp{rp}"].iloc[event_start_indices] = True
+    # Go through each date and add activations
+    for rp in RP_LIST:
+        rp_val = df_return_period.loc[rp, station]
+        for percentile in FORECAST_PERCENTILE_LIST:
+            for event_type, leadtimes in LEADTIMES_BY_TRIGGER.items():
+                data[f'event_{event_type}_rp{rp}_p{percentile}'] = data.apply(lambda row: 
+                                          (row[[f'forecast_lt{leadtime}_p{percentile}' for leadtime in leadtimes]] >= rp_val).any(),
+                                          axis=1)
+    df_station_dict[station] = data
+data
+```
+
+### Get TP, FP, and FN
+
+```python
+# Find TP, FP and FN
+
+def get_consecutive_groups(x: pd.Series, n=2):
+    # Find indices whre False / True change. Skip the first entry of the array because it always changes there
+    l = np.where(x.shift() != x)[0][1:]
+    # Convert array to pairs
+    return [l[i:i+n] for i in range(0, len(l), n)]
+
+
+def get_station_stats(df_station_dict, event_var, rp=MAIN_RP):
+
+    days_before_buffer = 0 # Event can occur at the earliest on the same day as the trigger
+    days_after_buffer = 30 # How many days the true event can occur after the GloFAS event
+    df_station_stats = pd.DataFrame(columns=['station', 'TP', 'FP', 'FN', 'event_type', 'percentile'])
+
+    for station in STATIONS:
+        df_station = df_station_dict[station]
+        if event_var == 'event_danger':
+            df_station = df_station.dropna(subset=['water_level'])
+        df_true_events = df_station[df_station[event_var]][[event_var]]
+        rp_val = df_return_period.loc[rp, station]
+        for event_type, leadtimes in LEADTIMES_BY_TRIGGER.items():
+            for percentile in FORECAST_PERCENTILE_LIST:
+                glofas_event_indices = get_consecutive_groups(df_station[f"event_{event_type}_rp{rp}_p{percentile}"])
+                df_true_events['detections'] = 0
+                TP = 0
+                FP = 0    
+                for (glofas_event_start_index, glofas_event_end_index) in glofas_event_indices:
+                    glofas_event = df_station.index[glofas_event_start_index]
+                    # Check the maximum lead time 
+                    max_leadtime = -1
+                    for leadtime in leadtimes:
+                        if df_station.loc[glofas_event, f'forecast_lt{leadtime}_p{percentile}'] >= rp_val:
+                            max_leadtime = leadtime
+                    if max_leadtime == -1:
+                        print('Something went wrong!')
+                    # Check if any events are around that date
+                    days_offset = (df_true_events.index - glofas_event) /  np.timedelta64(1, 'D')
+                    # Add the max leadtime to the days after buffer
+                    detected = (days_offset >= -1 * days_before_buffer) & (days_offset <= days_after_buffer + max_leadtime)
```

Comment:
Also an interesting way to look at detections that is slightly different than the approach that I've been taking in my work. This looks good to me but it would be great to have a quick chat on it together so that we're generally taking the same approach :)

Reply:
For sure, I also realized that there are a couple of ways that you could define TPs depending on what you want. I will ping you and we can plan a chat